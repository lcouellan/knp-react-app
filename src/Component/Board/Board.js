import React from 'react'

// Board :: Props -> React.Component
export default ({
  addCard,
  isLoading,
  cards
}) =>
  <main className="myBoard">
    {cards.map((value, key) =>
      <div key={key} className="card">
        <h3 className="card-title">{value.title}</h3>
      </div>
    )}

    {isLoading
      ? <div className="board loader">We are adding a new card...</div>
      : <button className="board add-card" onClick={addCard}>Add another card</button>
    }
  </main>
