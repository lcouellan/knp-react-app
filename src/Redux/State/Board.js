import { createReducer } from '../../utils'
import { always } from 'ramda'

// INITIAL_STATE :: State
export const INITIAL_STATE = {
  cards: [],
  isBlankCardVisible: false,
  blankCardTitle: ''
}

export const SAVE_CARD = '@knp/Board/SaveCard'
export const SET_BLANK_CARD_TITLE = '@knp/Board/SetBlankCardTitle'
export const TOGGLE_BLANK_CARD = '@knp/Board/ToggleBlankCard'
export const REMOVE_BLANK_CARD = '@knp/Board/RemoveBlankCard'


// saveCard :: () -> Action SAVE_CARD
export const saveCard = always({ type: SAVE_CARD })

// setBlankCardTitle :: () -> Action SET_BLANK_CARD_TITLE
export const setBlankCardTitle = cardTitle => ({
  type: SET_BLANK_CARD_TITLE,
  cardTitle
})

// toggleBlankCard :: () -> Action TOGGLE_BLANK_CARD
export const toggleBlankCard = always({ type: TOGGLE_BLANK_CARD })

// removeBlankCard :: () -> Action REMOVE_BLANK_CARD
export const removeBlankCard = always({ type: REMOVE_BLANK_CARD })

// Board :: (State, Action *) -> State
export default createReducer(INITIAL_STATE, {
  [SAVE_CARD]: state => ({
    ...state,
    blankCardTitle: '',
    isBlankCardVisible: false,
    cards: [
      ...state.cards,
      {
        title: state.blankCardTitle
      }
    ]
  }),
  [SET_BLANK_CARD_TITLE]: (state, { cardTitle }) => ({
    ...state,
    blankCardTitle: cardTitle
  }),
  [TOGGLE_BLANK_CARD]: state => ({
    ...state,
    isBlankCardVisible: true
  }),
  [REMOVE_BLANK_CARD]: state => ({
    ...state,
    isBlankCardVisible: false,
    blankCardTitle: ''
  })
})
