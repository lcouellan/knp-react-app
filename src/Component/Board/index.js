import { connect } from 'react-redux'
import { showBlankCard, saveCard, setBlankCardTitle } from '../../Redux/State/Board'
import { compose } from 'ramda'
import Board from './Board'

// mapStateToProps :: State -> Props
const mapStateToProps = state => ({
  cards: state.Board.cards,
  isBlankCardVisible: state.Board.isBlankCardVisible,
  cardTitle: state.Board.blankCardTitle
})

// mapDispatchToProps :: (Action * -> State) -> Props
const mapDispatchToProps = dispatch => ({
  addCard: compose(dispatch, showBlankCard),
  saveCard: compose(dispatch, saveCard),
  setBlankCardTitle: compose(dispatch, setBlankCardTitle)
})

export default connect(mapStateToProps, mapDispatchToProps)(Board)
