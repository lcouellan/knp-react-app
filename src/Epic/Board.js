import { combineEpics, ofType } from 'redux-observable'
import * as Board from '../Redux/State/Board'
import { map } from 'rxjs/operators'

// addCardEpic :: Epic -> Observable Action Board.ADD_CARD
export const addCardEpic = (action$) => action$.pipe(
  ofType(Board.ADD_CARD),
  map(() => ({
    title: 'My title'
  })),
  map(Board.receiveCard)
)

export default combineEpics(
  addCardEpic
)
