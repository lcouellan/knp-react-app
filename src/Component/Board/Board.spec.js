import React from 'react'
import Board from './Board'
import TestRenderer, { act } from 'react-test-renderer'

describe('Component :: Board :: Board', () => {
  it('loads when a card is creating', () => {
    let inspector

    act(() => {
      inspector = TestRenderer.create(
        <Board
          isLoading={true}
          cards={[]}
        />
      )
    })

    expect(inspector).toMatchSnapshot()
  })

  it('displays the cards and add "add another card button" when it is not creating any card', () => {
    let inspector

    act(() => {
      inspector = TestRenderer.create(
        <Board
          isLoading={false}
          cards={[{ title: 'My title' }]}
        />
      )
    })

    expect(inspector).toMatchSnapshot()
  })
})
