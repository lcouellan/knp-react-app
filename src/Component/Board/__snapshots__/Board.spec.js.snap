// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component :: Board :: Board displays the cards and add "add another card button" when it is not creating any card 1`] = `
<main
  className="myBoard"
>
  <div
    className="card"
  >
    <h3
      className="card-title"
    >
      My title
    </h3>
  </div>
  <button
    className="board add-card"
  >
    Add another card
  </button>
</main>
`;

exports[`Component :: Board :: Board loads when a card is creating 1`] = `
<main
  className="myBoard"
>
  <div
    className="board loader"
  >
    We are adding a new card...
  </div>
</main>
`;
